src/metric/auc.cu

```diff
@@ -0,0 +1,547 @@
+#include <cub/cub.cuh>
+#include <cassert>
+#include <limits>
+#include <memory>
+#include <utility>
+#include <tuple>
+
+#include "rabit/rabit.h"
+#include "xgboost/span.h"
+#include "xgboost/data.h"
+#include "auc.h"
+#include "../common/device_helpers.cuh"
+#include "../common/ranking_utils.cuh"
+
+namespace xgboost {
+namespace metric {
+namespace {
+template <typename T>
+class Discard : public thrust::discard_iterator<T>  {
+ public:
+  using value_type = T;  // NOLINT
+};
+
+struct GetWeightOp {
+  common::Span<float const> weights;
+  common::Span<size_t const> sorted_idx;
+
+  __device__ float operator()(size_t i) const {
+    return weights.empty() ? 1.0f : weights[sorted_idx[i]];
+  }
+};
+}  // namespace
+
+/**
+ * A cache to GPU data to avoid reallocating memory.
+ */
+struct DeviceAUCCache {
+  // Pair of FP/TP
+  using Pair = thrust::pair<float, float>;
+  // index sorted by prediction value
+  dh::device_vector<size_t> sorted_idx;
+  // track FP/TP for computation on trapesoid area
+  dh::device_vector<Pair> fptp;
+  // track negative/positive for computation on trapesoid area
+  dh::device_vector<Pair> neg_pos;
+  // index of unique prediction values.
+  dh::device_vector<size_t> unique_idx;
+  // p^T: transposed prediction matrix, used by MultiClassAUC
+  dh::device_vector<float> predts_t;
+  std::unique_ptr<dh::AllReducer> reducer;
+
+  void Init(common::Span<float const> predts, bool is_multi, int32_t device) {
+    if (sorted_idx.size() != predts.size()) {
+      sorted_idx.resize(predts.size());
+      fptp.resize(sorted_idx.size());
+      unique_idx.resize(sorted_idx.size());
+      neg_pos.resize(sorted_idx.size());
+      if (is_multi) {
+        predts_t.resize(sorted_idx.size());
+        reducer.reset(new dh::AllReducer);
+        reducer->Init(rabit::GetRank());
+      }
+    }
+  }
+};
+
+/**
+ * The GPU implementation uses same calculation as CPU with a few more steps to distribute
+ * work across threads:
+ *
+ * - Run scan to obtain TP/FP values, which are right coordinates of trapesoid.
+ * - Find distinct prediction values and get the corresponding negative/postive value,
+ *   which are left coordinates of trapesoid.
+ * - Reduce the scan array into 1 AUC value.
+ */
+std::tuple<float, float, float>
+GPUBinaryAUC(common::Span<float const> predts, MetaInfo const &info,
+             int32_t device, std::shared_ptr<DeviceAUCCache> *p_cache) {
+  auto& cache = *p_cache;
+  if (!cache) {
+    cache.reset(new DeviceAUCCache);
+  }
+  cache->Init(predts, true, device);
+
+  auto labels = info.labels_.ConstDeviceSpan();
+  auto weights = info.weights_.ConstDeviceSpan();
+  dh::safe_cuda(cudaSetDevice(device));
+
+  CHECK(!labels.empty());
+  CHECK_EQ(labels.size(), predts.size());
+
+  /**
+   * Create sorted index for each class
+   */
+  auto d_sorted_idx = dh::ToSpan(cache->sorted_idx);
+  dh::ArgSort<false>(predts, d_sorted_idx);
+
+  /**
+   * Linear scan
+   */
+  dh::caching_device_vector<float> d_auc(1);
```

Comment:
If you specify the fill value as 0 in the caching_device_vector constructor, it should be safely initialised.